### ISSUE TITLE
Allow patterns matching in conditional expression condition.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/language#2664]
We allow `if (e case p when e) ...` as statements and elements, but not in as expressions, where conditional branching uses the `?`/`:` operator.

I suggest we allow it, but require parentheses, so:
```dart
(e case p when e) ? ... : ....
```
becomes a valid expression.

It will not introduce any new expressive power, since `[if (e case p when e) ... else ...].first` does precisely the same thing, just with more memory overhead. It completes the orthogonality of the feature, by allowing it in every branch construct. 

(Can also be allowed in a `while`-loop condition, with a `true` scope covering the body of the `while`.)

**Comment by munificent** [comment]
This seems reasonable to me too, but I agree with marking it "patterns-later" so we can get a feel for how the pattern matching features we've already specced fit into the ecosystem. I'm always worried about designing too far ahead of ourselves before we get real hands on feedback from how users adopt a feature. (I realize the irony of me saying that given the huge surface area of patterns already...)

**Comment by FaFre** [comment]
I came across this issue today and was surprised this wasn't working as expected.

```dart
_baseUrl.replace(
    queryParameters: {
      ...
      if (document case SharedUrl(uri: final url)) 'url': url,
    },
    fragment: (document case SharedText(text: final text)) ? text : null, //doesn't work :(
  );
```

It would be great to see support for this :)

For now I think a switch is a good workaround, but doesn't feel very natural:
```dart
fragment: switch (document) {
      SharedText(text: final text) => text,
      _ => null,
    }
```

**Comment by tatumizer** [comment]
> It will not introduce any new expressive power, since [if (e case p when e) ... else ...].first does precisely the same thing 

A nicer form of conditional  expression would be  (if (e case p when e) ... else...) , which is style-consistent with collection if, and remains valid if you later introduce if expression with no parentheses

**Comment by FMorschel** [comment]
This would be perfect for cases like:

```dart
class C({var int? f});

void foo(C? c) {
  var v = (c?.f case var f?) ? f : 0;
}
```

Instead of doing something like:

```dart
void foo(C? c) {
  var v = 0;
  if (c?.f case var f?) v = f;
  // OR
  int v;
  if (c?.f case var f?) {
    v = f;
  } else {
    v = 0;
  }
}
```

Do we have any plans on making this happen soon? Considering we already have similar syntax I don't suppose this would be too hard?

And just to link a similar (or duplicate?) issue with more upvotes:

- https://github.com/dart-lang/language/issues/3059

**Comment by Wdestroier** [comment]
Isn't this code:
```dart
var v = (c?.f case var f?) ? f : 0;
```
Equivalent to the following?
```dart
var v = c?.f ?? 0;
```

**Comment by FMorschel** [comment]
Yes, it was a poor example. But the gist is the same, using patterns at that position and giving an alternate option (when the value is not null this is more useful as you just said hahaha).

**Comment by ykmnkmi** [comment]
```dart
while (token.next case var next?) {
  // ...
  token = next;
}
```

**Comment by lrhn** [comment]
Yeah, I catch myself writing `while (... case...)` occasionally. It feels like it should work. 
Especially if I convert an `if (...) { ...}` to a `while`.

And you can always write `while (true) { if (... case ...) {...} else {break;}}` for the same effect.
